src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs

```diff
@@ -772,13 +774,28 @@ protected DynamoViewModel(StartConfiguration startConfiguration)
             MLDataPipelineExtension = model.ExtensionManager.Extensions.OfType<DynamoMLDataPipelineExtension>().FirstOrDefault();
         }
 
+
+        private void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
+        {
+            try
+            {
+                var crashData = new CrashErrorReportArgs(e.Exception);
+                Model?.Logger?.LogError($"Unobserved task exception: {crashData.Details}");
+                Analytics.TrackException(e.Exception, true);
+            }
+            catch
+            { }
+        }
+
         private void CurrentDispatcher_UnhandledException(object sender, DispatcherUnhandledExceptionEventArgs e)
```

Comment:
I think it might be useful for host integrators in the same process. If we have an unhandled exception CurrentDispatcher_UnhandledException handler is the only way to gracefully close Dynamo and not terminate Revit too.
It may not be as useful for sandbox scenarios.

Reply:
I think the dispatcher unhandled exception handler in D4R that currently exists today suffers from the same issue where the stack is not unwound in some cases. If that happens the handler tries to shutdown Dynamo but then the code in the remaining call stack continues to execute causing further exceptions to be thrown down the line that are uncaught. At least that's what I saw while debugging.

Reply:
So I don't think the one in DynamoRevit is a great solution either.

Reply:
Yes, the dispatcher handler sometimes resumes execution of code after handling exceptions and that can lead to more exceptions which will bubble up to the host app handlers. (multiple CER reports)

There is some value in the dispatcher handler, specifically that we can isolate a crash in Dynamo and leave host app running.

We could consider all Dynamo uncaught exceptions to be fatal to the entire process (but that would change existing behavior)

Reply:
> There is some value in the dispatcher handler, specifically that we can isolate a crash in Dynamo and leave host app running.
This will be true only if this handler is triggered in those cases where the stack is unwound, correct?

Reply:
```
Dispatcher L0: PackageManagerCLient constructor => throw;
WPF callstack........................................
Dispatcher L1:  Try { invoke.CodeThatWEventuallyThrows;
} catch { 
Dispatcher L2: if (!handler.Handled) 
Dispatcher L3:    {throw;}
}
Dispatcher L4: Native Window messages
Wpf callstack.....................................
Dispatcher L5: Window.SHow()
DynamoView L6: packageManagerClient.Show();
DynamoView L7: packageManagerClient.SomeAPI;
Wpf callstack.....................................
Native Revit windows callstack....................
```
When Dispatcher L0 is executed and throws, the classtack is unwound until the first try catch (inside the DIspatcher)
Then dispatcher handlers are executed to check if exception should be silenced (and continue execution) or if exception should be rethrown
IF Handled, code keeps executing from Dispatcher L4: to Dispatcher L7. 
But in between (because initialization failed) Window.CLose is called which makes the packageManagerClient null before it gets to line DynamoView L7
This looks like an edge case and is not desireable.

Reply:
The second exception would be solved (in this case at least) if we put the CrashGracefully call into a Dispatcher.BeginInvoke() ...so that dynamo is not closed before the current dispatcher task has had a chance to finish execution. Of course this would not be a guarantee...depends on what the continuation executes..could still crash.

Reply:
Is this the call stack for the monocle view extension causing the crash?

Reply:
This is what's happening:
```
DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.PerformShutdownSequence(Dynamo.ViewModels.DynamoViewModel.ShutdownParams shutdownParams) Line 3522	C#
 	DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.Exit(object allowCancel) Line 2817	C#
 	DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.CrashGracefully(System.Exception ex) Line 796	C#
 	DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.CurrentDispatcher_UnhandledException(object sender, System.Windows.Threading.DispatcherUnhandledExceptionEventArgs e) Line 783	C#
	...
	...
>	DynamoCoreWpf.dll!Dynamo.Controls.DynamoView.DynamoViewModelRequestShowPackageManager(object s, System.EventArgs e) Line 2298	C# (packagemanagerwindow.show => throw)
 	DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.OnRequestPackageManagerDialog(object sender, System.EventArgs e) Line 31	C#
 	DynamoCoreWpf.dll!Dynamo.ViewModels.DynamoViewModel.ShowPackageManager(object parameters) Line 2317	C#
 	DynamoCoreWpf.dll!Dynamo.UI.Commands.DelegateCommand.Execute(object parameter) Line 38	C#
```
Once Dynamo shutsdown, then callstack unwinds to the line after line 2298 in `DynamoViewModelRequestShowPackageManager` and resumes execution.  

This line is `packagemanagerwindow.isloaded` and throws another exception ⚠️ ⚠️  (null reference exception) because of resuming execution, which in turn is because of _**callstack not unwinding**_ all the way down to the host equivalent of `dynamoRevit.rundynamo()`. My concern was this second exception but maybe it's not a big deal as you said since it's an edge case (maybe). And maybe this second exception can be handled by an `AppDomain.CurrentDomain.UnhandledException` handler on the host (DynamoRevit) side?

Reply:
I want to make sure we are in agreement that the **callstack not unwinding all the way down to the host equivalent of Dynamosandbox.RunApplication** is the whole point of the dispatcher handler `Handled=true`. If we use Handled=false, then the exception will cause a stack unwind until it hits the RunApplication try catch, or the process will be terminated by the clr.
**My concern was this second exception** - We can bypass this second exception if we run the Dynamo shutdown code in a Dispatcher.BeginInvoke task. Or if we do not, Revit will catch the exception. But we will generate 2 CER reports and the process will terminate

Reply:
I'm not sure I understand. I don't think we are in agreement.